README.md
# gb-dumper [![Build Status](https://travis-ci.org/gernd/gb-dumper.svg?branch=master)](https://travis-ci.org/gernd/gb-dumper)

Simple command line tool for displaying information about GB roms

[Changelog](CHANGELOG.md)

## Usage

The only parameter is the gb ROM file to be dumped.

    $ lein run [path-to-rom-file]

## License

Distributed under the [MIT License](LICENSE)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
## [0.2.0] - 2018-05-15
### Added
Parsing of ROM name

Parsing of GB color ROM indication
### Fixed
Skipping region between interrupt addresses and start opcodes

## [0.1.0] - 2018-04-11
### Changed
License change to MIT license

### Removed
Unused docs

Unused sections in Readme

### Fixed
Usage description

Unpacking of interrupt addresses

URL in project file

## [0.0.1] - 2018-04-09
### Added
First version parsing bytes up to the scrolling start logo.

[Unreleased]: https://github.com/gernd/gb-dumper/compare/v0.2.0...master
[0.2.0]: https://github.com/gernd/gb-dumper/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/gernd/gb-dumper/compare/v0.0.1...v0.1.0
[0.0.1]: https://github.com/gernd/gb-dumper/compare/ca6130150d7728ef7190c84a013ebc45b8e383e1...v0.0.1


test/gb_dumper/core_test.clj
(ns gb-dumper.core-test
  (:require [clojure.test :refer :all]
            [gb-dumper.analyze :refer :all])
  (:import (java.nio ByteBuffer)))


(defn prepare-test-data
  "Prepares test ROM data and returns it as byte[]"
  []
  (let [test-rom-size 8
        byte-buffer (ByteBuffer/allocate test-rom-size)
        buf (byte-array test-rom-size)]
    (.put byte-buffer (.byteValue 0x12))                    ; rst$00
    (.put byte-buffer (.byteValue 0x33))                    ; rst$08
    (.put byte-buffer (.byteValue 0x00))                    ; rst$10
    (.put byte-buffer (.byteValue 0x12))                    ; rst$18
    (.put byte-buffer (.byteValue 0x13))                    ; rst$20
    (.put byte-buffer (.byteValue 0x03))                    ; rst$28
    (.put byte-buffer (.byteValue 0x24))                    ; rst$30
    (.put byte-buffer (.byteValue 0x42))                    ; rst$38
    (.flip byte-buffer)
    (.get byte-buffer buf)
    buf))

(deftest test-unpack-rom-data-restart-addresses-correct
  (testing "Checks that the restart adresses are correctly read from the ROM"
    (let [test-rom-data (prepare-test-data)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data))
          (is (= 0x12 (get unpacked-rom-data :rst$00)))
          (is (= 0x33 (get unpacked-rom-data :rst$08)))
          (is (= 0x00 (get unpacked-rom-data :rst$10)))
          (is (= 0x12 (get unpacked-rom-data :rst$18)))
          (is (= 0x13 (get unpacked-rom-data :rst$20)))
          (is (= 0x03 (get unpacked-rom-data :rst$28)))
          (is (= 0x24 (get unpacked-rom-data :rst$30)))
          (is (= 0x42 (get unpacked-rom-data :rst$38)))))))


test/gb_dumper/analyze_test.clj
(ns gb-dumper.analyze-test
  (:require [clojure.test :refer :all]
            [gb-dumper.analyze :refer :all])
  (:import (java.nio ByteBuffer)))

(def valid-scrolling-nintendo-graphic
  (byte-array
    [
     0xCE 0xED 0x66 0x66 0xCC 0x0D 0x00 0x0B 0x03 0x73 0x00 0x83 0x00 0x0C 0x00 0x0D
     0x00 0x08 0x11 0x1F 0x88 0x89 0x00 0x0E 0xDC 0xCC 0x6E 0xE6 0xDD 0xDD 0xD9 0x99
     0xBB 0xBB 0x67 0x63 0x6E 0x0E 0xEC 0xCC 0xDD 0xDC 0x99 0x9F 0xBB 0xB9 0x33 0x3E
     ]))

(def invalid-scrolling-nintendo-graphic
  (byte-array
    [
     0xCE 0xED 0x66 0x66 0xCC 0x0D 0x00 0x0B 0x03 0x73 0x00 0x83 0x00 0x0C 0x00 0x0D
     0x00 0xFF 0x11 0x1F 0x88 0x89 0x00 0x0E 0xDC 0xCC 0x6E 0xE6 0xDD 0xDD 0xD9 0x99
     0xBB 0xBB 0x67 0x63 0xDD 0x0E 0xEC 0xCC 0xEE 0xDC 0x99 0x01 0xBB 0xB9 0x33 0x3E
     ]))

(def rom-name-16-bytes "name of the ROM!")

(def rom-name-less-than-16-bytes "RPG")

(def test-rom-size (+ 277 (count valid-scrolling-nintendo-graphic)))

(defn add-reset-addresses-interrupts-start-opcodes
  "Adds reset addresses, interrupts and start opcodes to the given byte buffer"
  [byte-buffer]
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x12)))      ; rst$00
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x33)))      ; rst$08
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x00)))      ; rst$10
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x12)))      ; rst$18
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x13)))      ; rst$20
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x03)))      ; rst$28
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x24)))      ; rst$30
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x99)))      ; rst$38
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x01)))      ; vertical blank interrupt
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x02)))      ; lcdc state interrupt
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x03)))      ; timer overflow interrupt
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x04)))      ; serial transfer complete interrupt
  (dotimes [n 8] (.put byte-buffer (.byteValue 0x05)))      ; high to low interrupt
  (dotimes [n 152] (.put byte-buffer (.byteValue 0x05)))    ; space up to 0x100
  (.put byte-buffer (.byteValue 0x12))                      ; code execution point (4 bytes)
  (.put byte-buffer (.byteValue 0x13))                      ; code execution point (4 bytes)
  (.put byte-buffer (.byteValue 0x21))                      ; code execution point (4 bytes)
  (.put byte-buffer (.byteValue 0x70)))                     ; code execution point (4 bytes)

(defn prepare-valid-test-data-rom-name-16-bytes
  "Prepares valid test ROM data with a name of exactly 16 bytes and returns it as byte[]"
  []
  (let [byte-buffer (ByteBuffer/allocate test-rom-size)
        buf (byte-array test-rom-size)]
    (add-reset-addresses-interrupts-start-opcodes byte-buffer)
    (.put byte-buffer valid-scrolling-nintendo-graphic 0 (count valid-scrolling-nintendo-graphic))
    (.put byte-buffer (.getBytes rom-name-16-bytes) 0 16)
    (.put byte-buffer (.byteValue 0x80))
    (.flip byte-buffer)
    (.get byte-buffer buf)
    buf))

(defn prepare-valid-test-data-is-gb-color-rom
  "Prepares valid test ROM data indicating that it is a GB color ROM and returns it as byte[]"
  []
  (let [byte-buffer (ByteBuffer/allocate test-rom-size)
        buf (byte-array test-rom-size)]
    (add-reset-addresses-interrupts-start-opcodes byte-buffer)
    (.put byte-buffer valid-scrolling-nintendo-graphic 0 (count valid-scrolling-nintendo-graphic))
    (.put byte-buffer (.getBytes rom-name-16-bytes) 0 16)
    (.put byte-buffer (.byteValue 0x80))
    (.flip byte-buffer)
    (.get byte-buffer buf)
    buf))

(defn prepare-valid-test-data-is-not-gb-color-rom
  "Prepares valid test ROM data indicating that it is not a GB color ROM and returns it as byte[]"
  []
  (let [byte-buffer (ByteBuffer/allocate test-rom-size)
        buf (byte-array test-rom-size)]
    (add-reset-addresses-interrupts-start-opcodes byte-buffer)
    (.put byte-buffer valid-scrolling-nintendo-graphic 0 (count valid-scrolling-nintendo-graphic))
    (.put byte-buffer (.getBytes rom-name-16-bytes) 0 16)
    (.put byte-buffer (.byteValue 0x34))
    (.flip byte-buffer)
    (.get byte-buffer buf)
    buf))

(defn prepare-valid-test-data-rom-name-less-than-16-bytes
  "Prepares valid test ROM data and returns it as byte[]"
  []
  (let [byte-buffer (ByteBuffer/allocate test-rom-size)
        buf (byte-array test-rom-size)
        number-of-padding-bytes (- 16 (count rom-name-less-than-16-bytes))]
    (add-reset-addresses-interrupts-start-opcodes byte-buffer)
    (.put byte-buffer valid-scrolling-nintendo-graphic 0 (count valid-scrolling-nintendo-graphic))
    (.put byte-buffer (.getBytes rom-name-less-than-16-bytes) 0 (count rom-name-less-than-16-bytes))
    (dotimes [n number-of-padding-bytes] (.put byte-buffer (.byteValue 0x00)))
    (.put byte-buffer (.byteValue 0x80))
    (.flip byte-buffer)
    (.get byte-buffer buf)
    buf))

(defn prepare-invalid-test-data
  "Prepares invalid test ROM data and returns it as byte[]"
  []
  (let [byte-buffer (ByteBuffer/allocate test-rom-size)
        buf (byte-array test-rom-size)]
    (add-reset-addresses-interrupts-start-opcodes byte-buffer)
    (.put byte-buffer invalid-scrolling-nintendo-graphic 0 (count valid-scrolling-nintendo-graphic))
    (.put byte-buffer (.getBytes rom-name-16-bytes) 0 16)
    (.put byte-buffer (.byteValue 0x80))
    (.flip byte-buffer)
    (.get byte-buffer buf)
    buf))

(defn is-same-byte
  "Checks that the byte value of two expressions is the same"
  [first-byte second-byte]
  (is (= (.byteValue first-byte) (.byteValue second-byte))))

(defn is-same-byte-vector
  "Checks if two vectors contain the same bytes"
  [first-vector second-vector]
  (is (= (seq (map #(.byteValue %) first-vector))
         (seq (map #(.byteValue %) second-vector)))))

(deftest test-unpack-rom-data-restart-addresses
  (testing "Checks that the restart adresses are correctly read from the ROM"
    (let [test-rom-data (prepare-valid-test-data-rom-name-16-bytes)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data) " unpacked ROM data is " unpacked-rom-data)
          (is-same-byte-vector [0x12 0x12 0x12 0x12 0x12 0x12 0x12 0x12] (get unpacked-rom-data :rst$00))
          (is-same-byte-vector [0x33 0x33 0x33 0x33 0x33 0x33 0x33 0x33] (get unpacked-rom-data :rst$08))
          (is-same-byte-vector [0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00] (get unpacked-rom-data :rst$10))
          (is-same-byte-vector [0x12 0x12 0x12 0x12 0x12 0x12 0x12 0x12] (get unpacked-rom-data :rst$18))
          (is-same-byte-vector [0x13 0x13 0x13 0x13 0x13 0x13 0x13 0x13] (get unpacked-rom-data :rst$20))
          (is-same-byte-vector [0x03 0x03 0x03 0x03 0x03 0x03 0x03 0x03] (get unpacked-rom-data :rst$28))
          (is-same-byte-vector [0x24 0x24 0x24 0x24 0x24 0x24 0x24 0x24] (get unpacked-rom-data :rst$30))
          (is-same-byte-vector [0x99 0x99 0x99 0x99 0x99 0x99 0x99 0x99] (get unpacked-rom-data :rst$38))))))

(deftest test-unpack-rom-data-interrupt-addresses
  (testing "Checks that the interrupt adresses are correctly read from the ROM"
    (let [test-rom-data (prepare-valid-test-data-rom-name-16-bytes)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data) " unpacked ROM data is " unpacked-rom-data)
          (is-same-byte-vector [0x01 0x01 0x01 0x01 0x01 0x01 0x01 0x01] (get unpacked-rom-data :vertical-blank-interrupt))
          (is-same-byte-vector [0x02 0x02 0x02 0x02 0x02 0x02 0x02 0x02] (get unpacked-rom-data :lcdc-status-interrupt))
          (is-same-byte-vector [0x03 0x03 0x03 0x03 0x03 0x03 0x03 0x03] (get unpacked-rom-data :timer-overflow-interrupt))
          (is-same-byte-vector [0x04 0x04 0x04 0x04 0x04 0x04 0x04 0x04] (get unpacked-rom-data :serial-transfer-completion-interrupt))
          (is-same-byte-vector [0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05] (get unpacked-rom-data :high-to-low-interrupt))))))

(deftest test-unpack-rom-data-start-opcodes
  (testing "Checks that the start opcodes are correctly read from the ROM"
    (let [test-rom-data (prepare-valid-test-data-rom-name-16-bytes)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data) " unpacked ROM data is " unpacked-rom-data)
          (is-same-byte 0x12 (nth (get unpacked-rom-data :start-opcodes) 0))
          (is-same-byte 0x13 (nth (get unpacked-rom-data :start-opcodes) 1))
          (is-same-byte 0x21 (nth (get unpacked-rom-data :start-opcodes) 2))
          (is-same-byte 0x70 (nth (get unpacked-rom-data :start-opcodes) 3))))))

(deftest test-unpack-rom-data-valid-scrolling-logo
  (testing "Checks that a valid scrolling logo is correctly read from the ROM and detected as valid"
    (let [test-rom-data (prepare-valid-test-data-rom-name-16-bytes)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data) " unpacked ROM data is " unpacked-rom-data)
          (is (= (seq valid-scrolling-nintendo-graphic) (seq (get unpacked-rom-data :logo))))
          (is (get unpacked-rom-data :logo-is-valid))))))

(deftest test-unpack-rom-data-invalid-scrolling-logo
  (testing "Checks that a invalid scrolling logo is correctly read from the ROM and detected as invalid"
    (let [test-rom-data (prepare-invalid-test-data)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data) " unpacked ROM data is " unpacked-rom-data)
          (is (= (seq invalid-scrolling-nintendo-graphic) (seq (get unpacked-rom-data :logo))))
          (is (not (get unpacked-rom-data :logo-is-valid)))))))

(deftest test-unpack-rom-data-name-with-16-bytes
  (testing "Checks that a name of 16 bytes is correctly read from the ROM"
    (let [test-rom-data (prepare-valid-test-data-rom-name-16-bytes)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data) " unpacked ROM data is " unpacked-rom-data)
          (is (= rom-name-16-bytes (get unpacked-rom-data :name)))))))

(deftest test-unpack-rom-data-name-with-less-than-16-bytes
  (testing "Checks that a name of less than 16 bytes is correctly read from the ROM"
    (let [test-rom-data (prepare-valid-test-data-rom-name-less-than-16-bytes)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data) " unpacked ROM data is " unpacked-rom-data)
          (is (= rom-name-less-than-16-bytes (get unpacked-rom-data :name)))))))

(deftest test-unpack-rom-data-is-gb-color-rom
  (testing "Checks that a GB color ROM is correctly parsed"
    (let [test-rom-data (prepare-valid-test-data-is-gb-color-rom)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data) " unpacked ROM data is " unpacked-rom-data)
          (is (get unpacked-rom-data :is-gb-color-rom))))))

(deftest test-unpack-rom-data-is-not-gb-color-rom
  (testing "Checks that a non GB color ROM is correctly parsed"
    (let [test-rom-data (prepare-valid-test-data-is-not-gb-color-rom)
          unpacked-rom-data (unpack-rom-data test-rom-data)]
      (do (println "Test ROM data is" (to-hex-string test-rom-data) " unpacked ROM data is " unpacked-rom-data)
          (is not (get unpacked-rom-data :is-gb-color-rom))))))


src/gb_dumper/core.clj
(ns gb-dumper.core
  (:gen-class))

(require '(clojure.java [io :as io]))
(use 'gb-dumper.analyze)

(defn print-usage
  "Prints usage information on howto use the GB ROM Analyzer"
  []
  (println "Usage: executable <PATH-TO-GB-ROM>"))

(defn read-rom
  "Reads the given ROM into a byte array"
  [file]
  (with-open [xin (io/input-stream file)
              xout (java.io.ByteArrayOutputStream.)]
    (io/copy xin xout)
    (.toByteArray xout)))

(defn -main
  "Analyzes the GB ROM file given as the first parameter "
  [& args]
  (if (not (= 1 (count args)))
    (print-usage)
    (let [file-name (first args)]
      (do (println "Analyzing GB Rom File " file-name)
          (let [rom-binary-data (read-rom file-name)
                unpacked-rom-data (unpack-rom-data rom-binary-data)]
            (println unpacked-rom-data))))))


src/gb_dumper/analyze.clj
(ns gb-dumper.analyze
  (:import (java.nio ByteBuffer)
           (java.nio.charset Charset)))

(def expected-logo-graphic
  (byte-array
    [
     0xCE 0xED 0x66 0x66 0xCC 0x0D 0x00 0x0B 0x03 0x73 0x00 0x83 0x00 0x0C 0x00 0x0D
     0x00 0x08 0x11 0x1F 0x88 0x89 0x00 0x0E 0xDC 0xCC 0x6E 0xE6 0xDD 0xDD 0xD9 0x99
     0xBB 0xBB 0x67 0x63 0x6E 0x0E 0xEC 0xCC 0xDD 0xDC 0x99 0x9F 0xBB 0xB9 0x33 0x3E
     ]))

(defn to-hex-string
  "Convert a seq of bytes into a hex encoded string"
  [bytes]
  (apply str "0x" (for [byte bytes] (format "%02x" byte))))

(defn unpack-bytes
  "unpacks the given number of bytes from a ByteBuffer as vector of bytes"
  [byte-buffer number-of-bytes]
  (let [byte-arr (byte-array number-of-bytes)]
    (.get byte-buffer byte-arr)
    byte-arr))

(defn unpack-start-opcodes
  "unpacks the start opcodes from the ROM given as ByteBuffer"
  [rom-data-buffer]
  (unpack-bytes rom-data-buffer 4))

(defn unpack-logo-graphic
  "unpacks the start logo graphic from the ROM given as ByteBuffer"
  [rom-data-buffer]
  (unpack-bytes rom-data-buffer 48))

(defn unpack-eight-bytes
  "unpacks eight bytes from the ROM given as ByteBuffer"
  [rom-data-buffer]
  (unpack-bytes rom-data-buffer 8))

(defn unpack-name
  "unpacks the name for the ROM from the given ByteBuffer"
  [rom-data-buffer]
  (let [name-bytes (unpack-bytes rom-data-buffer 16)
        name-without-zero-bytes (byte-array (filter #(not (= % 0x00)) (vec name-bytes)))]
    (new String (bytes name-without-zero-bytes) (Charset/forName "US-ASCII"))))

(defn unpack-rom-data
  "unpacks the gb rom data given as byte array"
  [rom-data]
  (let [byte-buffer (ByteBuffer/allocate (count rom-data))]
    (.put byte-buffer rom-data 0 (count rom-data))          ; fill byte buffer with rom data
    (.flip byte-buffer)                                     ; reset pointer for reading
    (let [
          ; restart addresses
          rst$00 (unpack-eight-bytes byte-buffer)
          rst$08 (unpack-eight-bytes byte-buffer)
          rst$10 (unpack-eight-bytes byte-buffer)
          rst$18 (unpack-eight-bytes byte-buffer)
          rst$20 (unpack-eight-bytes byte-buffer)
          rst$28 (unpack-eight-bytes byte-buffer)
          rst$30 (unpack-eight-bytes byte-buffer)
          rst$38 (unpack-eight-bytes byte-buffer)
          ; interrupt addresses
          vertical-blank-interrupt (unpack-eight-bytes byte-buffer)
          lcdc-status-interrupt (unpack-eight-bytes byte-buffer)
          timer-overflow-interrupt (unpack-eight-bytes byte-buffer)
          serial-transfer-completion-interrupt (unpack-eight-bytes byte-buffer)
          high-to-low-interrupt (unpack-eight-bytes byte-buffer)
          ; code execution start opcodes start at 0x100, skip to there
          unused (.get byte-buffer (byte-array 152))
          ; code execution start opcodes
          code-execution-start-opcodes (unpack-start-opcodes byte-buffer)
          ; scrolling start logo
          start-logo (unpack-logo-graphic byte-buffer)
          ; name of the ROM
          name (unpack-name byte-buffer)
          ; flag indicating whether this ROM is a GB color ROM
          is-gb-color-rom (= (.byteValue 0x80) (.get byte-buffer))
          ]
      {
       :rst$00                               rst$00
       :rst$08                               rst$08
       :rst$10                               rst$10
       :rst$18                               rst$18
       :rst$20                               rst$20
       :rst$28                               rst$28
       :rst$30                               rst$30
       :rst$38                               rst$38
       :vertical-blank-interrupt             vertical-blank-interrupt
       :lcdc-status-interrupt                lcdc-status-interrupt
       :timer-overflow-interrupt             timer-overflow-interrupt
       :serial-transfer-completion-interrupt serial-transfer-completion-interrupt
       :high-to-low-interrupt                high-to-low-interrupt
       :start-opcodes                        code-execution-start-opcodes
       :logo                                 start-logo
       :logo-is-valid                        (= (seq expected-logo-graphic) (seq start-logo))
       :name                                 name
       :is-gb-color-rom                      is-gb-color-rom
       })))
